_posts/2020-01-21-fault-tolerance-4-0.adoc

```diff
@@ -0,0 +1,147 @@
+It has been originally based on {uri-hystrix}[Hystrix], the Netflix library for latency and fault tolerance for distributed systems.
+Lately, we realized that Hystrix is not the best fit for implementing MicroProfile Fault Tolerance, because Hystrix meshes all fault tolerance concerns into one, while MicroProfile Fault Tolerance describes mostly a layered architecture.
+The MicroProfile Fault Tolerance specification also requires certain features that Hystrix intentionally doesn't provide; most importantly, the ability to interrupt threads that Hystrix itself didn't create.
+
+We have been talking on and off about removing Hystrix and using our own implementation of core fault tolerance strategies, one that would be better suited as a base for implementing MicroProfile Fault Tolerance, but only a few months ago did we finally start working on it.
+Thanks to heroic work of Michał Szynkiewicz, the new implementation didn't take long to pass the entire MicroProfile Fault Tolerance TCK.
+We have already released a first version of this new implementation, called 4.0.0.
+Of course, that's not the end; in fact, it's just a beginning.
+We should now have a better foundation to experiment with, and already have several ideas that we would like to explore and eventually propose to MicroProfile Fault Tolerance.
+
+In this article, we'll take a brief look at how the new implementation works inside, what do you need to know as an integrator, and what are our plans for the future.
+
+== Under the hood
+
+First, let's take a brief look at the core.
+It is based on the idea of `FaultToleranceStrategy`, which is an interface that looks like this:
+
+[source,java]
+----
+interface FaultToleranceStrategy<V> {
+    V apply(InvocationContext<V> ctx) throws Exception;
+}
+----
+
+The `InvocationContext` is a `Callable` that represents the method invocation guarded by this fault tolerance strategy.
+The fault tolerance strategy does its work around `ctx.call()`.
+For example, it can catch exceptions, invoke `ctx.call()` multiple times, invoke something else, etc.
+
+In the codebase, you can find implementations of this interface for all the strategies required by MicroProfile Fault Tolerance: retry, fallback, timeout, circuit breaker or bulkhead.
+Asynchronous invocation, delegated to a thread pool, is of course also supported.
+When multiple fault tolerance strategies are supposed to be used to guard one method, they can form a chain.
+
+In fact, all the implementations expect to be used in a chain, which is why they all take another `FaultToleranceStrategy` to which they delegate.
+The ultimate `ctx.call()` invocation is done by a special fault tolerance strategy which is called, well, `Invocation`.
+
+As an example, let's consider this method:
+
+[source,java]
+----
+@Fallback(...)
+@Retry(...)
+@Timeout(...)
+public void doSomething() {
+    ...
+}
+----
+
+The chain of fault tolerance strategies will look roughly like this:
+
+[source]
```

Comment:
Purely for my curiosity, does the ordering of the annotations on the method dictate the chaining?

If so, does that have any implications for the ordering? ie. that a specific order is recommended to chain them in an appropriate manner? Should `Fallback` always be the first annotation so it's the outermost chain?

Reply:
The order in the chain is specified by MP FT and the order in which the annotations are defined does not modify it.

Reply:
awesome, thanks.

Just a coincidence that the annotations matched up then

Reply:
I reordered the annotations and added an explanatory sentence. Good point, I didn't even realize that the order was the same :-)